File: Python/0031. 整数拆分.py
# If you need to import additional packages or classes, please import here.
def func():
    # please define the python3 input here. 
    # For example: a,b = map(int, input().strip().split())
    # please finish the function body here.
    # please define the python3 output here. For example: print().
    while True:
        try:
            n = int(input())
            arr = [[0]*(n+1) for _ in range(n+1)]
            
            for i in range(1,n+1):
                for j in range(1,n+1):
                    if i==1 or j==1:
                        arr[i][j] = 1
                    elif i<j:
                        arr[i][j] = arr[i][i]
                    elif i==j:
                        arr[i][j] = 1+arr[i][j-1]
                    else:
                        arr[i][j] = arr[i-j][j] + arr[i][j-1]
            print(arr[n][n])
                
            
            
            
        except EOFError:
            break
    
if __name__ == "__main__":
    func()

File: C++/0031. 整数拆分.cpp
// we have defined the necessary header files here for this problem.
// If additional header files are needed in your program, please import here.
#include<vector>
int main()
{  
  // please define the C++ input here. For example: int a,b; cin>>a>>b;;  
  // please finish the function body here.  
  // please define the C++ output here. For example:cout<<____<<endl; 

    int n;
    while(cin>>n){
        vector<vector<int>> arr(n+1,vector<int>(n+1,0));
        for(int i=1;i<=n;++i){
            for(int j=1;j<=n;++j){
                if(i==1||j==1){
                    arr[i][j] = 1;
                }
                else if(i<j){
                    arr[i][j] = arr[i][i];
                }
                else if(i==j){
                    arr[i][j] = 1+arr[i][j-1];
                }
                else{
                    arr[i][j] = arr[i-j][j] + arr[i][j-1];
                }
            }
        }
        cout<<arr[n][n]<<endl;
    }
    return 0;
}
